### PR TITLE
cmake: Add -gdwarf-4 at link time for gcc

### LINKED ISSUE #50106: ram_report stopped working with zephyr-sdk 0.15
**Describe the bug**
Since Zephyr requires 0.15 of the SDK now, I upgraded to that version, and then the ram report broke because of an outdated pyelftools, so I updated that to 0.29.

**To Reproduce**
For ChromeOS EC, we have a publicly visible CI system running on gitlab.com. You can see this failure at https://gitlab.com/zephyr-ec/ec/-/pipelines/635841472 but here is the excerpt:

```
ninja -C ${b} ram_report >"${BUILD_DIR}/${PROJECT}/output/${bdir}_ram_report.txt"

Traceback (most recent call last):
  File "/builds/zephyr-ec/ec/zephyr/main/scripts/footprint/size_report", line 803, in <module>
    main()
  File "/builds/zephyr-ec/ec/zephyr/main/scripts/footprint/size_report", line 779, in main
    do_simple_name_matching(elf, symbol_dict, processed)
  File "/builds/zephyr-ec/ec/zephyr/main/scripts/footprint/size_report", line 317, in do_simple_name_matching
    symbol = match_symbol_address(symlist, die,
  File "/builds/zephyr-ec/ec/zephyr/main/scripts/footprint/size_report", line 114, in match_symbol_address
    low, high = get_die_mapped_address(die, parser, dwarfinfo)
  File "/builds/zephyr-ec/ec/zephyr/main/scripts/footprint/size_report", line 84, in get_die_mapped_address
    loc = parser.parse_from_attribute(loc_attr, die.cu['version'])
  File "/home/uboot/.local/lib/python3.8/site-packages/elftools/dwarf/locationlists.py", line 307, in parse_from_attribute
    return self.location_lists.get_location_list_at_offset(
  File "/home/uboot/.local/lib/python3.8/site-packages/elftools/dwarf/locationlists.py", line 48, in get_location_list_at_offset
    raise DWARFError("For this binary, \"die\" needs to be provided")
elftools.common.exceptions.DWARFError: For this binary, "die" needs to be provided
```

**Impact**
Just an annoyance.

**Environment (please complete the following information):**
 - OS: Linux docker
 - Toolchain: Zephyr sdk 0.15
 - Commit SHA: Anything after 30af51fc5e0ccdc2b612f9912726a7211c35f13e

### PR DESCRIPTION
**Pull request by jeremybettis**
Add a linker option for gcc to force dwarf symbols to be version 4.

Fixes #50106
Related to #35707

Signed-off-by: Jeremy Bettis <jbettis@google.com>

### CODE REVIEW

File: cmake/linker/ld/gcc/linker_flags.cmake
Comment (stephanosio) [inline cmake/linker/ld/gcc/linker_flags.cmake:14]: All "Zephyr" executables should already be using DWARF v4 because of 72c2f182bc4170ccffa13a52ad99eb5d0f74adc2.

Are you adding a target other than `zephyr_interface` in your build system?

Comment (jeremybettis) [reply cmake/linker/ld/gcc/linker_flags.cmake:14]: I looked for all instances of TARGET in our cmake files. We use zephyr_library_named() to create some other named libraries. Here is an example of what that looks like

```
zephyr_library_named(ec_shim)
get_property(APP_LTO_PROPERTY TARGET app PROPERTY INTERPROCEDURAL_OPTIMIZATION)
set_property(TARGET ec_shim PROPERTY INTERPROCEDURAL_OPTIMIZATION ${APP_LTO_PROPERTY})
set_property(TARGET ec_shim APPEND PROPERTY COMPILE_DEFINITIONS __REQUIRE_ZEPHYR_GPIOS__)
```

In any case, I can't find a dwarf-5 symbol in any .obj file, but they still appear in the zephyr-pre0.elf file. And all gcc compile commands include -gdwarf-4. The only commands that don't have them are the link steps, and adding it here fixes the problems and I an run the ram_report and rom_report again.

Comment (stephanosio) [reply cmake/linker/ld/gcc/linker_flags.cmake:14]: Fair enough. Using `add_link_options`, although not ideal, seems to the simplest workaround for that problem.

No major complaint about it since this is only a temporary workaround until the DWARF v5 format is properly supported by Zephyr and its tooling.

**Review by stephanosio — APPROVED** [review]

**Review by dcpleung — APPROVED** [review]

File: cmake/linker/ld/gcc/linker_flags.cmake
Comment (tejlmand) [inline cmake/linker/ld/gcc/linker_flags.cmake:14]: a late arrival comment, as I see this has already been merged.

This file is not intended to add symbols to any targets, it is intended to setup linker properties, and then the build system at another location decides if / when and where this property should be used.

So it should have been something like:
```
set_property(TARGET linker PROPERTY debug -gdwarf-4)
```

and then let other part of build system decide when to use it.

But as this is temporary solution, then no reason to make an update to this.

On the interesting part, Pyelftool 0.28 includes initial support for dwarf-5:
> + Version 0.28 (2022.02.03)
> ....
>     - Initial DWARF v5 support (#363 with several follow-ups)
https://github.com/eliben/pyelftools/pull/363

**Comment by stephanosio** [comment]
Please do not merge PRs without the assignee's approval (especially those marked with "Maintainer" label).

File: cmake/linker/ld/gcc/linker_flags.cmake
Comment (stephanosio) [reply cmake/linker/ld/gcc/linker_flags.cmake:14]: > On the interesting part, Pyelftool 0.28 includes initial support for dwarf-5:

https://github.com/zephyrproject-rtos/zephyr/pull/46609. It turns out something is still broken even with pyelftools 0.29.

Comment (tejlmand) [reply cmake/linker/ld/gcc/linker_flags.cmake:14]: > #46609. It turns out something is still broken even with pyelftools 0.29.

yes, from initial support in 0.28 and more fixes in 0.29 I do assume there is still some way to go.
Just wanted to raise awareness, but seems you already tried it out :+1: .
Guess I was still trying to get through my backlog at that time, cause I can't remember #46608 :smile:

**Comment by mmahadevan108** [comment]
> Please do not merge PRs without the assignee's approval (especially those marked with "Maintainer" label).

Noted, will look for this.

**Pull request merged**
